### PR TITLE
Remove redundant dag_id index on log table

### PR DESCRIPTION
**Pull request by dstandish**
This index is redundant since there's another one with dag_id as leading column.

### CODE REVIEW

**Review by vincbeck — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]

File: airflow/migrations/versions/0033_3_0_0_remove_redundant_index.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+revision = "16cbcb1c8c36"
+down_revision = "522625f6d606"
+branch_labels = None
+depends_on = None
+airflow_version = "3.0.0"
+
+
+def upgrade():
+    """Apply Remove redundant index."""
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (ephraimbuddy) [inline airflow/migrations/versions/0033_3_0_0_remove_redundant_index.py]:
```suggestion
```

**Comment by kaxil** [comment]
I was reading about what the recommendation is around Multi-column indexes vs single -- and looks like it isn't as simple as I had initially thought that the multi-column index with a leading column can remove the need of a separate index in all cases!

From https://www.postgresql.org/docs/current/indexes-bitmap-scans.html:
>In all but the simplest applications, there are various combinations of indexes that might be useful, and the database developer must make trade-offs to decide which indexes to provide. Sometimes multicolumn indexes are best, but sometimes it's better to create separate indexes and rely on the index-combination feature. For example, if your workload includes a mix of queries that sometimes involve only column x, sometimes only column y, and sometimes both columns, you might choose to create two separate indexes on x and y, relying on index combination to process the queries that use both columns. You could also create a multicolumn index on (x, y). This index would typically be more efficient than index combination for queries involving both columns, but as discussed in [Section 11.3](https://www.postgresql.org/docs/current/indexes-multicolumn.html), it would be almost useless for queries involving only y, so it should not be the only index. A combination of the multicolumn index and a separate index on y would serve reasonably well. For queries involving only x, the multicolumn index could be used, though it would be larger and hence slower than an index on x alone. The last alternative is to create all three indexes, but this is probably only reasonable if the table is searched much more often than it is updated and all three types of query are common. If one of the types of query is much less common than the others, you'd probably settle for creating just the two indexes that best match the common types.

**Comment by jedcunningham** [comment]
Interesting, TIL. Probably still a good general approach though, but yeah in some cases it could make sense to keep them around.

This table specifically is so write heavy, wouldn't be worth it :)

**Comment by kaxil** [comment]
> Interesting, TIL. Probably still a good general approach though, but yeah in some cases it could make sense to keep them around.
> 
> This table specifically is so write heavy, wouldn't be worth it :)

Yeah, I was fixing some migrations issues and found similar indexes and I thought let me remove them and got curious what the recommendation -- and backtracked on mixing it with some of the changes I will have in a PR

**Comment by dstandish** [comment]
> I was reading about what the recommendation is around Multi-column indexes vs single -- and looks like it isn't as simple as I had initially thought that the multi-column index with a leading column can remove the need of a separate "single-col" index in all cases!

I did not really see anything in there that suggested that it's not true.

If you have index on x, y, the queries on x are covered by the index.  That's true.  Now, sure, an index on x alone be smaller. 
 Yes. But it remains to be shown what difference that makes.

**Comment by kaxil** [comment]
This sentence is the key there: "**_For queries involving only x, the multicolumn index could be used, though it would be larger and hence slower than an index on x alone._** "

How much slower? That depends on the dataset and access pattern.

In above, a multi-column index on (x, y) technically covers queries on x alone but the additional storage and scanning costs can impact performance, especially for high-traffic columns. A single-column index on x is smaller and offers a narrower scan range, which often improves I/O and cache usage, making a measurable difference in efficiency for large datasets. Again how much difference depends on the dataset and query patterns

**Comment by dstandish** [comment]
right yeah i saw that.  i think that the default choice would be, don't add the redundant index unless you had a compelling reason.  that being a little bigger would make it measurably slower isn't surprising but that doesn't mean it's worth adding.  i think the side proposing to add the redundant index would bear the burden to demonstrate the need for it.  the fact that one index can serve multiple query patterns is a great thing and we should take advantage of that where we can i think.

**Comment by dstandish** [comment]
implication of that being, if you see this sort of thing, probably best to remove, don't you think?

**Comment by kaxil** [comment]
Agreed

**Pull request merged**